src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -84,61 +97,343 @@
+     */
+    public static byte[] getStreamAsByteArray(final InputStream stream)
+    {
+        if (stream == null)
+        {
+            Log.getLogger().info("Structure.getStreamAsByteArray: stream is null this should not happen");
+            return new byte[0];
+        }
+        try
+        {
+            ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+
+            int nRead;
+            byte[] data = new byte[BUFFER_SIZE];
+
+            while ((nRead = stream.read(data, 0, data.length)) != -1)
             {
-                Log.getLogger().warn("Couldn't find any structure with this name anywhere", e);
+                buffer.write(data, 0, nRead);
             }
+            return buffer.toByteArray();
         }
+        catch (@NotNull IOException e)
+        {
+            Log.getLogger().trace(e);
+        }
+        return new byte[0];
+    }
 
-        if (inputstream == null)
+    /**
+     * Calculate the MD5 hash for a template from an inputstream.
+     *
+     * @param stream to which we want the MD5 hash
+     * @return the MD5 hash string or null
+     */
+    public static String calculateMD5(final InputStream stream)
+    {
+        if (stream == null)
         {
-            return;
+            Log.getLogger().error("Structure.calculateMD5: stream is null, this should not happen");
+            return null;
         }
+        return calculateMD5(getStreamAsByteArray(stream));
+    }
 
+    /**
+     * Calculate the MD5 hash of a byte array
+     *
+     * @param bytes array
+     * @return the MD5 hash string or null
+     */
+    public static String calculateMD5(final byte[] bytes)
+    {
         try
         {
-            this.template = readTemplateFromStream(inputstream);
+            final MessageDigest md = MessageDigest.getInstance("MD5");
+            return DatatypeConverter.printHexBinary(md.digest(bytes));
         }
-        catch (final IOException e)
+        catch (@NotNull NoSuchAlgorithmException e)
         {
-            Log.getLogger().warn(String.format("Failed to load template %s", structureName), e);
+            Log.getLogger().trace(e);
         }
-        finally
+
+        return null;
+    }
+
+    /**
+     * Compare the md5 from the structure with an other md5 hash.
+     *
+     * @param otherMD5 to compare with
+     * @return whether the otherMD5 match, return false if md5 is null
+     */
+    public boolean isCorrectMD5(final String otherMD5)
+    {
+        Log.getLogger().info("isCorrectMD5: md5:" + md5 + " other:" + otherMD5);
+        if (md5 == null || otherMD5 == null)
+        {
+            return false;
+        }
+        return md5.compareTo(otherMD5) == 0;
+    }
+
+    public static byte[] compress(final byte[] data)
+    {
+        final ByteArrayOutputStream byteStream = new ByteArrayOutputStream(data.length);
+        try (GZIPOutputStream zipStream = new GZIPOutputStream(byteStream))
+        {
+             zipStream.write(data);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Make this line start at column 13. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AIndentationCheck)